### ISSUE TITLE
Add _bytearray_release functions; optimize bytearray->bytes conversion

### ISSUE THREAD

**Issue by tekknolagi** [tekknolagi/skybison#173]
T56760468:
The encoders and _io code uses pattern like this:
a = bytearray()
... add stuff to a ...
return bytes(a)
We could optimize things by adding a _bytearray_release(x) function that takes a bytearray x and then returns the contents of the bytearray as a bytes object and resets the bytearray to be empty.

Benefits:
Avoid some of the type.__call__ machinery (which is still slower than we hope, but we will optimize that too in the future).
We can start playing tricks and add a function that takes the internal buffer of the bytearray, changes the layout to RawLargeBytes (if it is big enough) and resets the object length. (We have to add a function to shrink objects though, we likely need a way to mark the unused bytes after shrinking somehow).
It's probably helpful to add this even without the shrinking logic. If you do so please be sure to file a separate task for the shrinking before closing this!

Created-Date: 2019/11/01 09:38:18
Blocks: T53256673, T56760319, T55028419
